Update Cloudflare/GA analytics tools to latest design/behavior

blackjackkent opened the pull request:
#### Changes proposed in this Pull Request

* This PR refactors the GA/Cloudflare Analytics cards on Marketing -> Traffic to match the designs listed in pbAPfg-1fn-p2. It also adds logic by which GA will be disabled when a new Cloudflare tracking ID gets set, and vice versa.

#### Testing instructions

* Check out this diff to your local Calypso environment.
* Check out D56767-code to your sandbox.
* Navigate to Tools->Marketing->Traffic on a free site account.
* Verify that the Cloudflare and Google Analytics tool options are listed as requiring a premium or higher account.

![image](https://user-images.githubusercontent.com/13437011/107697627-155a1580-6c79-11eb-91b4-9e4161e3d987.png)

* Navigate to Tools->Marketing->Traffic on a premium account.
*  Verify that you can enable the Cloudflare analytics feature.
* Paste the following snippet into the text field:
```
<!-- Cloudflare Web Analytics -->
<script defer 
    src='https://static.cloudflareinsights.com/beacon.min.js' 
    data-cf-beacon='{"token": "716d4d275266427cb78934b7919e73e6"}'>
</script>
<!-- End Cloudflare Web Analytics -->
```
* Verify that when tabbing out of the field (or otherwise blurring it) this snippet is condensed to only the token value (i.e. `716d4d275266427cb78934b7919e73e6`)
* After saving this update, verify that, when visiting the site, you can see a Cloudflare analytics block in the footer code:
<img width="1462" alt="Screen Shot 2021-02-03 at 2 19 47 PM" src="https://user-images.githubusercontent.com/13437011/106804409-f9c48e80-662a-11eb-8566-17bd2bdc83f4.png">
* Return to Tools->Marketing->Traffic. Verify that you can enable the Google Analytics plugin as well.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 428-gh-Automattic/dotcom-manage

blackjackkent committed:
Cloudflare analytics card redesign

blackjackkent committed:
File reorganization

blackjackkent committed:
Refactoring google form

blackjackkent committed:
Adding confirmation modal

a8ci18n commented on client/my-sites/site-settings/analytics/form-cloudflare-analytics.js:
```diff
@@ -0,0 +1,286 @@
+				{ isModalVisible && (
+					<AnalyticsToggleConfirmationDialog
+						productToDisable="Google Analytics"
+						productToEnable="Cloudflare"
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Cloudflare CDN' )` *ES Score: 6*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/cloudflare-analytics-save-settings)

blackjackkent commented:
There is currently a bug in this which is causing the logic that triggers the confirmation modal to appear to only work the first time it is run on a given page load. Toggling back again to the other service doesn't trigger it again unless you reload the page first.

I'm pretty baffled as to why this is at the moment so if anyone reviewing this has any idea, please let me know. (It's kind of an edge case though so I'm not sure if it counts as a blocker to merge.)

gwwar commented on client/my-sites/site-settings/analytics/form-cloudflare-analytics.js:
```diff
@@ -0,0 +1,286 @@
+		if ( fields?.cloudflare_analytics?.code ) {
+			setIsCloudflareEnabled( true );
+		}
+	}, [ fields ] );
```
Maybe try ```[ fields?.cloudflare_analytics?.code ]``` instead here, if we can't trust the array ref to be updated on any state changes

kwight commented:
If we're stuck on certain behaviour, are there any parts that can be broken out into separate PRs that would be closer to merge? Like the card refactoring?

blackjackkent committed:
Removing Cloudflare/GA exclusivity and updating settings field name

blackjackkent commented:
Updated based on discussion in p1612908228011400-slack-C01HV13S295 which removed the complexity of trying to make GA/Cloudflare usage exclusive to each other. Also updated site setting name based on discussion in https://github.com/Automattic/jetpack/pull/18771

a8ci18n commented on client/my-sites/site-settings/analytics/form-cloudflare-analytics.js:
```diff
@@ -0,0 +1,263 @@
+					isSaving={ isSavingSettings }
+					onButtonClick={ handleSubmitForm }
+					showButton
+					title={ translate( 'Cloudflare' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Cloudflare CDN' )` *ES Score: 6*

blackjackkent committed:
Fixing google logo

blackjackkent commented:
Updated testing instructions based on changes to D56767-code.

kwight commented:
I tested this with D56767-code and it worked well, except for the case of pasting in a snippet. When pasting, the form field displays the correct extracted ID, but the settings API call is wrapping the value in an array. Manually entering any valid value sends the correct value (not wrapped in an array).

| Pasted | Manual |
| --- | --- |
| <img width="727" alt="Screen Shot 2021-02-17 at 1 05 28 PM" src="https://user-images.githubusercontent.com/349751/108268556-8be78f00-7121-11eb-8e63-7d35bd000068.png"> | <img width="583" alt="Screen Shot 2021-02-17 at 1 06 24 PM" src="https://user-images.githubusercontent.com/349751/108268638-ab7eb780-7121-11eb-9665-e9a39f1850b8.png"> |

mreishus commented:
* On a free site, I see the upsell banners linking me to purchase a plan
* On a premium site, I can set my CloudFlare and GA ids
  * Also seeing the same Array bug as Kirk
```
<script defer
	src='https://static.cloudflareinsights.com/beacon.min.js'
	data-cf-beacon='{"token": "Array"}'>
``` 
* Are modals still included?
  * I saw a comment about modals not always working, but I wasn't able to get any modals to display when turning on and off GA/Cloudflare analytics.

blackjackkent commented:
> Are modals still included?

@mreishus See the discussion in p1612908228011400-slack-C01HV13S295 - we decided to remove that requirement. :)

Good catch (with @kwight ) on the array issue; I will fix that today.

blackjackkent committed:
Fixing regex bug

blackjackkent commented:
Pushed a fix to use the proper key from the regex result when parsing a pasted JS snippet.

kwight commented:
Looks good – I can see the snippet being added with the key now. I also saw some CORS errors – not sure if they're a problem (that can be followup).

kwight reviewed (approved):
🚢

blackjackkent committed:
Retriggering Github checks

blackjackkent committed:
Fixing test import

blackjackkent committed:
Fixing regex parsing

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5535772

Thank you @blackjackkent for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5535772).
